### PR TITLE
ARROW-14345: [C++] Implement streaming reads

### PR DESCRIPTION
**Pull request by coryan**

### CODE REVIEW

**Review by emkornfield — CHANGES_REQUESTED** [review]
Mostly just questions about approach. on somethings and checking for consistency with other file system APIs

**Comment by ElenaHenderson** [comment]
Benchmark runs are scheduled for baseline = 4ac62d5527e5f1635764a62986589d694605d1d8 and contender = a8e1c81249aa2477fd0623e3cfd9d50ba2eba4fd. a8e1c81249aa2477fd0623e3cfd9d50ba2eba4fd is a master commit associated with this PR. Results will be available as each benchmark for each run completes.
Conbench compare runs links:
[Scheduled] [ec2-t3-xlarge-us-east-2](https://conbench.ursa.dev/compare/runs/4bb9abb6eb644c78a05cfd85f9dad605...80e025da88d44d6cb8c1152e8ea7d1a0/)
[Finished :arrow_down:1.54% :arrow_up:0.0%] [ursa-i9-9960x](https://conbench.ursa.dev/compare/runs/c8fb01f5f8b544649fad11ad9fc5b8fa...34807827b4c442cbbb4345a05e9508ac/)
[Finished :arrow_down:0.09% :arrow_up:0.0%] [ursa-thinkcentre-m75q](https://conbench.ursa.dev/compare/runs/455daa163640479aae0d68cfb9694920...cef4ad2e7e5145919eb54b40d29c4afa/)
Supported benchmarks:
ursa-i9-9960x: langs = Python, R, JavaScript
ursa-thinkcentre-m75q: langs = C++, Java
ec2-t3-xlarge-us-east-2: cloud = True

File: cpp/src/arrow/filesystem/gcsfs.cc [hunk 1]
```diff
@@ -58,9 +61,48 @@ struct GcsPath {
...
+    return stream_.gcount();
+  }
+
+  Result<std::shared_ptr<Buffer>> Read(int64_t nbytes) override {
+    ARROW_ASSIGN_OR_RAISE(auto buffer, arrow::AllocateResizableBuffer(nbytes));
+    stream_.read(reinterpret_cast<char*>(buffer->mutable_data()), nbytes);
+    if (!stream_.status().ok()) {
+      return internal::ToArrowStatus(stream_.status());
+    }
+    return arrow::SliceMutableBufferSafe(std::move(buffer), 0, stream_.gcount());
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs.cc:98]: Hmm. This is merely slicing the output buffer, but not resizing it. If the actual size is much smaller than `nbytes`, then there is a significant memory wastage.

**Comment by pitrou** [comment]
I just posted a belated comment on this PR. Sorry for being so late :-)

File: cpp/src/arrow/filesystem/gcsfs.cc (hunk 1 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc:98]: True. On the other hand, a short read from a large buffer could require a lot of copying.  I am not sure which is more likely to be a problem.  I am happy to make a change if you have some data or intuition as to which one is worse.

Comment (pitrou) [reply cpp/src/arrow/filesystem/gcsfs.cc:98]: Does GCS return short reads except for the EOF case?

Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc:98]: For all practical purposes, only the EOF case.  The exception would be some kind of transient error that is stubborn enough to exhaust the retry policy.

Comment (pitrou) [reply cpp/src/arrow/filesystem/gcsfs.cc:98]: Ok. In any case, other implementations (such as S3) do resize the buffer, and there doesn't seem to be a reason to act differently here.

Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc:98]: Created https://issues.apache.org/jira/browse/ARROW-14559 to track that change.

**Pull request closed without merging**